go/vt/vtgate/planbuilder/insert.go

```diff
@@ -17,20 +17,89 @@ import (
 
 // buildInsertPlan builds the route for an INSERT statement.
 func buildInsertPlan(ins *sqlparser.Insert, vschema VSchema) (*engine.Route, error) {
-	route := &engine.Route{
-		Query: generateQuery(ins),
-	}
-	var err error
-	route.Table, err = vschema.Find(string(ins.Table.Qualifier), string(ins.Table.Name))
+	table, err := vschema.Find(string(ins.Table.Qualifier), string(ins.Table.Name))
 	if err != nil {
 		return nil, err
 	}
-	route.Keyspace = route.Table.Keyspace
-	if !route.Keyspace.Sharded {
-		route.Opcode = engine.InsertUnsharded
-		return route, nil
+	if !table.Keyspace.Sharded {
+		return buildInsertUnshardedPlan(ins, table, vschema)
+	}
+	return buildInsertShardedPlan(ins, table)
+}
+
+func buildInsertUnshardedPlan(ins *sqlparser.Insert, table *vindexes.Table, vschema VSchema) (*engine.Route, error) {
+	rt := &engine.Route{
+		Opcode:   engine.InsertUnsharded,
+		Table:    table,
+		Keyspace: table.Keyspace,
+	}
+	var values sqlparser.Values
+	switch rows := ins.Rows.(type) {
+	case *sqlparser.Union:
+		return nil, errors.New("unsupported: union in insert")
+	case *sqlparser.Select:
+		bldr, err := processSelect(rows, vschema, nil)
+		if err != nil {
+			return nil, err
+		}
+		rb, ok := bldr.(*route)
+		if !ok {
+			return nil, errors.New("unsupported: complex join in insert")
+		}
+		if rb.ERoute.Keyspace.Name != rt.Keyspace.Name {
+			return nil, errors.New("unsupported: cross-shard select in insert")
+		}
+		if rt.Table.AutoIncrement != nil {
+			return nil, errors.New("unsupported: auto-inc and select in insert")
+		}
+		rt.Query = generateQuery(ins)
```

Comment:
This is mainly for readability, to make this block of code stand on its own. Otherwise, any changes to the outer block can affect this block and cause bugs.